SignalServiceKit/src/Messages/TSMessagesManager.m

```diff
@@ -347,134 +362,172 @@ - (void)processEnvelope:(OWSSignalServiceProtosEnvelope *)envelope
-                dispatch_async(dispatch_get_main_queue(), ^{
-                    [self processException:exception envelope:messageEnvelope];
-                    NSString *errorDescription =
-                        [NSString stringWithFormat:@"Exception while decrypting: %@", exception.description];
-                    NSError *error = OWSErrorWithCodeDescription(OWSErrorCodeFailedToDecryptMessage, errorDescription);
-                    completion(error);
-                });
-                return;
-            }
+    [self decryptEnvelope:envelope
+        messageTypeName:@"PreKey Bundle"
+        missingPayloadBlock:^{
+            OWSProdFail([OWSAnalyticsEvents messageManagerErrorPrekeyBundleEnvelopeHasNoContent]);
+        }
+        cipherMessageBlock:^(NSData *encryptedData) {
+            return [[PreKeyWhisperMessage alloc] initWithData:encryptedData];
+        }
+        successBlock:successBlock
+        failureBlock:failureBlock];
+}
+
+- (void)decryptEnvelope:(OWSSignalServiceProtosEnvelope *)envelope
+        messageTypeName:(NSString *)messageTypeName
+    missingPayloadBlock:(void (^_Nonnull)())missingPayloadBlock
+     cipherMessageBlock:(id<CipherMessage> (^_Nonnull)(NSData *))cipherMessageBlock
+           successBlock:(DecryptSuccessBlock)successBlock
+           failureBlock:(void (^)(NSError *_Nullable error))failureBlock
+{
+    OWSAssert(envelope);
+    OWSAssert(messageTypeName.length > 0);
+    OWSAssert(missingPayloadBlock);
+    OWSAssert(cipherMessageBlock);
+    OWSAssert(successBlock);
+    OWSAssert(failureBlock);
+
+    TSStorageManager *storageManager = [TSStorageManager sharedManager];
+    NSString *recipientId = envelope.source;
+    int deviceId = envelope.sourceDevice;
+
+    // DEPRECATED - Remove after all clients have been upgraded.
+    NSData *encryptedData = envelope.hasContent ? envelope.content : envelope.legacyMessage;
+    if (!encryptedData) {
+        missingPayloadBlock();
+        failureBlock(nil);
+        return;
+    }
 
+    dispatch_async([OWSDispatch sessionStoreQueue], ^{
+        @try {
+            id<CipherMessage> cipherMessage = cipherMessageBlock(encryptedData);
+            SessionCipher *cipher = [[SessionCipher alloc] initWithSessionStore:storageManager
+                                                                    preKeyStore:storageManager
+                                                              signedPreKeyStore:storageManager
+                                                               identityKeyStore:self.identityManager
+                                                                    recipientId:recipientId
+                                                                       deviceId:deviceId];
+
+            NSData *plaintextData = [[cipher decrypt:cipherMessage] removePadding];
+            successBlock(plaintextData);
+        } @catch (NSException *exception) {
             dispatch_async(dispatch_get_main_queue(), ^{
```

Comment:
I'd love to move more off the main thread here, but I'll take this on later to avoid rebase hell with the next PR. https://trello.com/c/5wutIgRv/1562-move-more-work-in-message-decrypter-and-processor-off-main-threadd